src/hotspot/share/opto/castnode.hpp

```diff
@@ -69,6 +69,27 @@ class ConstraintCastNode: public TypeNode {
   static Node* make_cast_for_type(Node* c, Node* in, const Type* type, DependencyType dependency);
 
   Node* optimize_integer_cast(PhaseGVN* phase, BasicType bt);
+
+  // Visit all non-cast uses of the node, bypassing ConstraintCasts.
+  // Pattern: this (-> ConstraintCast)* -> non_cast
+  // In other words: find all non_cast nodes such that
+  // non_cast->uncast() == this.
+  template <typename Callback>
+  static void visit_uncasted_uses(const Node* n, Callback callback) {
+    Unique_Node_List internals;
+    internals.push((Node*)n); // start traversal
```

Comment:
We should get rid of these casts with [JDK-8252694](https://bugs.openjdk.org/browse/JDK-8252694). I added a comment there.